src/Adapter/Product/Update/ProductIndexationUpdater.php

```diff
@@ -68,21 +78,23 @@ public function isVisibleOnSearch(Product $product): bool
 
     /**
      * @param Product $product
-     *
-     * @throws CannotUpdateProductException
-     * @throws CoreException
+     * @param ShopConstraint $shopConstraint
      */
-    public function updateIndexation(Product $product): void
+    public function updateIndexation(Product $product, ShopConstraint $shopConstraint): void
     {
         if (!$this->isSearchIndexationOn) {
             return;
         }
 
+        $this->contextStateManager->setShopContext($shopConstraint);
+
         if ($this->isVisibleOnSearch($product)) {
             $this->updateProductIndexes((int) $product->id);
         } else {
             $this->removeProductIndexes((int) $product->id);
         }
+
+        $this->contextStateManager->restorePreviousContext();
```

Comment:
Usage of the context state manager must always be protected with a `try/finally` block like this:
```php
try {
    if ($this->isVisibleOnSearch($product)) {
            $this->updateProductIndexes((int) $product->id);
        } else {
            $this->removeProductIndexes((int) $product->id);
        }
} finally {
    $this->contextStateManager->restorePreviousContext();
}
```

Reply:
But I think you could handle the state manager in the `updateProductIndexes` and `removeProductIndexes` To keep the context wrapping the closest to the code that needs it

Reply:
Ideally, we shouldn't be relying on the state manager, we should be able to provide a parameter to the indexation methods but seeing how it's handled it will be complicated to refacto it right now Maybe it's worth adding some comment to explain this and justify why we used the state manager